kythe/go/extractors/validation/validation.go

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2018 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package validation contains logic for verifing the contents of a kzip.
+package validation
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"kythe.io/kythe/go/platform/kzip"
+)
+
+// Harness encapsulates data for validation of a single kzip and repo.
+type Harness struct {
+	// Path to the kzip file to validate.
+	Kzip string
+	// Path to the root of the repo to compare the kzip against.
+	Repo string
+	// Language extensions to check, e.g. java, cpp, h.
+	Langs map[string]bool
+	// Optional file to output missing paths to.
+	MissingOutput *string
+}
+
+// Result is a simple container struct explaining the results of validation.
+type Result struct {
+	// Number of files in the kzip file matching one of Langs extensions.
+	FilesInKzip int
+	// Number of files in the repo matching one of Langs extensions.
+	FilesInRepo int
+	// Number of files in the repo that were not matched in the kzip.
+	Missing int
+	// A breakdown of the top missing files, by subdirectory.
+	TopMissing Coverage
+}
+
+// Coverage describes the missing subdirectories in a repo.
+type Coverage struct {
+	// The missing paths.
+	Paths []Dir
+}
+
+// The number of missing paths to print in output.
+const topMissingPathsLimit = 20
+const topMissingPathsDepth = 5
+
+// Dir describes a single missing subdirectory and how much is missing.
+type Dir struct {
+	// Path is a subdirectory from Repo root.
+	Path string
+	// Missing is the number of files in the Repo from the corresponding
+	// subdirectory that were missing from the Kzip.
+	Missing int
+}
+
+// Len is part of sort.Interface.
```

Comment:
Please mention what length it reports, too, e.g.

    // Len reports the number of paths missing from the archive.

or words to that effect.

Reply:
:+1: